dali/dfuXRefLib/XRefNodeManager.cpp

```diff
@@ -176,143 +143,167 @@ StringBuffer & CXRefNode::getLastModified(StringBuffer & str)
+        // File-based storage: load from file
+        try
         {
-            messagesBranch = m_XRefTree->addPropTree("Messages",createPTree());
-            commit();
+            StringBuffer filepath(xrefPath);
+            addPathSepChar(filepath).append(branchName).append(".xml");
+            MemoryBuffer xmlContent;
+            readXrefBranch(branchName, filepath.str(), xrefMaxFileSize, xmlContent); // NB: includes null terminator
+            size32_t len = xmlContent.length();
+            if (len)
+                buf.setBuffer(len, (char *)xmlContent.detach(), len-1); // NB: StringBuffer::setBuffer strLen (3rd param) includes the null terminator
+        }
+        catch (IException *e)
+        {
+            StringBuffer errMsg;
+            OWARNLOG("XRefNode: Failed to load %s from file: %s", branchName, e->errorMessage(errMsg).str());
+            e->Release();
         }
-        StringBuffer tmpbuf;
-        m_messages.set(messagesBranch);
     }
-    buf.clear();
-    MemoryBuffer data;
-    m_messages->getPropBin("data",data);
-    if (data.length())
+    else
     {
-        buf.append(data.length(),data.toByteArray());
+        // Dali-based storage: use "data" attribute
+        if (!cached.get())
+        {
+            IPropertyTree *branch = m_XRefTree->queryPropTree(branchName);
+            if (!branch)
+            {
+                branch = m_XRefTree->addPropTree(branchName, createPTree());
+                commit();
+            }
+            cached.set(branch);
+        }
+        MemoryBuffer data;
+        cached->getPropBin("data", data);
+        if (data.length())
+        {
+            buf.append(data.length(), data.toByteArray());
+        }
     }
     return buf;
 }
 
-void CXRefNode::deserializeMessages(IPropertyTree& inTree)
+void CXRefNode::deserializeBranch(const char *branchName, Owned<IPropertyTree> &cached, IPropertyTree &inTree)
 {
-    if(!m_messages.get())
+    const char *xrefPath = m_XRefTree->queryProp("@xrefPath");
+
+    if (xrefPath && *xrefPath)
     {
-        IPropertyTree* messagesBranch = m_XRefTree->queryPropTree("Messages");
-        if(messagesBranch == 0)
+        // File-based storage: save to file
+        try
         {
-            messagesBranch = m_XRefTree->addPropTree("Messages",createPTree());
-            commit();
-        }
-        StringBuffer tmpbuf;
-        m_messages.set(messagesBranch);
-    }
-    StringBuffer datastr;
-    toXML(&inTree,datastr);
-    m_messages->setPropBin("data",datastr.length(),(void*)datastr.str());
-}
+            StringBuffer filepath(xrefPath);
+            addPathSepChar(filepath).append(branchName).append(".xml");
 
-StringBuffer &CXRefNode::serializeDirectories(StringBuffer &buf)
-{
-    if(!m_directories.get())
-    {
-        IPropertyTree* directoriesBranch = m_XRefTree->queryPropTree("Directories");
-        if(directoriesBranch == 0)
+            StringBuffer datastr;
+            toXML(&inTree, datastr);
```

Comment:
Would it be better to write directly to a file - especially if the data is large?